packages/controllers/src/ranks.rs

```diff
@@ -0,0 +1,257 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use thiserror::Error;
+
+use cosmwasm_std::{
+    attr, Addr, CustomQuery, Deps, DepsMut, MessageInfo, Response, StdError, StdResult,
+};
+use cw_storage_plus::Map;
+
+/// Returned from RANKS.query_rank()
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct RankResponse {
+    pub address_rank: Option<u8>, //more than 255 ranks seems impractical
+}
+
+/// Errors returned from RANKS
+#[derive(Error, Debug, PartialEq)]
+pub enum RanksError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("Caller has no rank")]
+    NoRank {},
+
+    #[error("Caller rank too low")]
+    RankTooLow {},
+
+    #[error("Caller cannot change their own rank")]
+    CannotSetOwnRank {},
+}
+
+// state/logic
+// an IndexedMap might make sense later so that admins of a given level can be easily pulled
+pub struct Ranks<'a>(Map<'a, &'a str, u8>);
+
+// this is the core business logic we expose
+impl<'a> Ranks<'a> {
+    pub const fn new(namespace: &'a str) -> Self {
+        Ranks(Map::new(namespace))
+    }
+
+    pub fn set<Q: CustomQuery>(
+        &self,
+        deps: DepsMut<Q>,
+        address: Addr,
+        address_rank: u8,
+    ) -> StdResult<()> {
+        self.0
+            .save(deps.storage, &address.to_string(), &address_rank)
+    }
+
+    pub fn get<Q: CustomQuery>(&self, deps: Deps<Q>, address: Addr) -> StdResult<Option<u8>> {
+        match self.0.load(deps.storage, &address.to_string()) {
```

Comment:
is 
if let Some(val) = self.0.may_load(...)? {
Ok(Some(Val)
} else {
Ok(None)
}
?

Reply:
or just

Ok(self.0.may_load( .. ) ?)